cpp/code/basic_arrow.cc

```diff
@@ -63,3 +66,54 @@ arrow::Status ReturnNotOk() {
 TEST(BasicArrow, ReturnNotOkNoMacro) { ASSERT_OK(ReturnNotOkMacro()); }
 
 TEST(BasicArrow, ReturnNotOk) { ASSERT_OK(ReturnNotOk()); }
+
+class TypeCountVisitor : public arrow::TypeVisitor {
+ public:
+  uint64_t nested_count;
+  uint64_t non_nested_count;
+
+  template <typename T>
+  arrow::enable_if_not_nested<T, arrow::Status> Visit(const T&) {
```

Comment:
I think I must be doing this wrong, since I am getting NotImplemented for int8:

https://github.com/apache/arrow-cookbook/runs/5588880880?check_suite_focus=true#step:6:94

Any idea on why?

Reply:
Is it using the default implementation from the parent class instead?

Reply:
Because you're not overriding the actual base method, you're defining an unrelated template method which happens to have the same name :-)

Reply:
`VisitTypeInline` may allow you to use such a pattern, because it's not based on polymorphism.

Reply:
Ah so if I understand right, if you subclass `TypeVisitor` you can't use templates to implement the overriding methods.

Reply:
Indeed, because you must override the exact methods. Adding `override` at the end helps ensure you didn't make a mistake when overriding, btw.